### ISSUE TITLE
Query: Allow certain unmapped constant in shaper expression

### ISSUE THREAD

**Issue by smitpatel** [aspnet/EntityFrameworkCore#17048]
Fix for #13048 involves trying to infer the constant expression in client code which are not server correlated and may contain reference to enclosing class which in terms capture context causing memory leak.
The heuristic we are currently using is, constantExpression
- whose value is not null
- & which are not mapped by current provider (since we don't send constants to server during client eval).

The set of conditions also includes certain other cases here. While it may not be always accurate but probably can be improved. 2 Failing cases right now.
```C#
        [ConditionalTheory]
        [MemberData(nameof(IsAsyncData))]
        public virtual Task GroupBy_empty_key_Aggregate_Key(bool isAsync)
        {
            return AssertQuery<Order>(
                isAsync,
                os =>
                    os.GroupBy(
                            o => new
                            {
                            })
                        .Select(
                            g => new
                            {
                                g.Key,
                                Sum = g.Sum(o => o.OrderID)
                            }));
        }
```
Group by empty key and then projecting such key (it works if key is not used in projection). Here we receive constant of anonymous type. 
We added these pattern for dev expression (ref #11905)
But we cannot find if the anonymous constant references enclosing class or not.
See also #15712 If that is fixed then perhaps we can fix this case.

```C#
[ConditionalTheory]
        [MemberData(nameof(IsAsyncData))]
        public virtual async Task ToString_with_formatter_is_evaluated_on_the_client(bool isAsync)
        {
            await AssertQuery<Order>(
                isAsync,
                os => os.Where(o => o.OrderDate != null)
                    .Select(
                        o => new Order
                        {
                            ShipName = o.OrderID.ToString(new CultureInfo("en-US"))
                        }),
                e => e.ShipName);
        }
```
CultureInfo is not mapped type. In future we could allow certain types we know not coming from client code to be in tree as we can be certain it cannot contain enclosing class.

**Comment by ajcvickers** [comment]
Notes from triage: @smitpatel to look at fixing the first case, since this is important for control vendors. Beyond that we still want to revisit this post 3.0.

**Comment by smitpatel** [comment]
Both of the cases worked using fix for #15712. So closing this alongwith.
